### PR TITLE
Improve PHPDoc types support

### LINKED ISSUE #4913: Bug in NoSuperfluousPhpdocTags
#### The PHP version you are using (`$ php -v`):
=> 7.4 (running on github actions)

#### PHP CS Fixer version you are using (`$ php-cs-fixer -V`):
=> v2.16.2 Yellow Bird

#### The command you use to run PHP CS Fixer:
=> php-cs-fixer fix

#### The configuration file you are using, if any:
```php
<?php

$finder = PhpCsFixer\Finder::create()
    ->in(__DIR__.'/src')
    ->in(__DIR__.'/tests')
;

return PhpCsFixer\Config::create()
    ->setRiskyAllowed(true)
    ->setRules([
        '@PSR2' => true,
        '@Symfony' => true,
        'array_syntax' => ['syntax' => 'short'],
        'blank_line_after_opening_tag' => true,
        'concat_space' => ['spacing' => 'one'],
        'declare_strict_types' => false,
        'fully_qualified_strict_types' => true,
        'header_comment' => false,
        'lowercase_cast' => true,
        'is_null' => true,
        'magic_method_casing' => true,
        'modernize_types_casting' => true,
        'multiline_comment_opening_closing' => true,
        'native_constant_invocation' => true,
        'no_alias_functions' => true,
        'no_alternative_syntax' => true,
        'no_blank_lines_after_phpdoc' => true,
        'no_empty_comment' => true,
        'no_empty_phpdoc' => true,
        'no_extra_blank_lines' => true,
        'no_leading_import_slash' => true,
        'no_spaces_around_offset' => true,
        'no_superfluous_phpdoc_tags' => ['allow_mixed' => true],
        'no_trailing_comma_in_singleline_array' => true,
        'no_unset_cast' => true,
        'no_unused_imports' => true,
        'no_useless_else' => true,
        'no_useless_return' => true,
        'no_whitespace_in_blank_line' => true,
        'ordered_imports' => true,
        'ordered_interfaces' => true,
        'php_unit_mock' => ['target' => 'newest'],
        'php_unit_dedicate_assert_internal_type' => ['target' => 'newest'],
        'php_unit_expectation' => ['target' => 'newest'],
        'php_unit_no_expectation_annotation' => ['target' => 'newest'],
        // 'php_unit_test_annotation' => ['style' => 'prefix'],
        'php_unit_test_case_static_method_calls' => ['call_type' => 'self'],
        'php_unit_mock_short_will_return' => true,
        'phpdoc_align' => ['align' => 'vertical'],
        'phpdoc_line_span' => ['method' => 'multi', 'property' => 'multi'],
        'phpdoc_scalar' => true,
        'phpdoc_separation' => true,
        'phpdoc_single_line_var_spacing' => true,
        'phpdoc_trim' => true,
        'phpdoc_trim_consecutive_blank_line_separation' => true,
        'phpdoc_types' => true,
        'phpdoc_types_order' => ['null_adjustment' => 'always_last', 'sort_algorithm' => 'none'],
        'phpdoc_var_without_name' => true,
        'short_scalar_cast' => true,
        'single_line_throw' => true,
        'single_trait_insert_per_statement' => true,
        'standardize_not_equals' => true,
        'visibility_required' => true,
        'yoda_style' => true,
        'native_function_invocation' => ['include' => ['@compiler_optimized']],
        'compact_nullable_typehint' => true,
        'declare_equal_normalize' => ['space' => 'none'],
        'function_typehint_space' => true,
        'no_leading_namespace_whitespace' => true,
        'blank_line_before_statement' => true,
        'single_quote' => true,
        'phpdoc_to_comment' => false,
        'ordered_class_elements' => ['sortAlgorithm'=>'none'],
        'class_attributes_separation' => ['elements'=>['property', 'method']],
        'array_indentation' => true,
    ])
    ->setFinder($finder)
;

```

#### If applicable, please provide minimum samples of PHP code (as plain text, not screenshots):
 * before running PHP CS Fixer (no changes):
```php
class MyEnum {
  const ONE=1;
  const TWO=2;
}

class Foo {
    /**
     * @var null|MyEnum::*
     */	
    private $numbers;

    /**
     * @var list<MyEnum::*>	
     */	
    private $numbers;
}
```

This is supported by PSALM: https://psalm.dev/docs/annotating_code/typing_in_psalm/#specifying-stringint-options-aka-enums

I dont expect these comments to be removed. Version v2.16.1 did not remove them. Version v2.16.1 does remove these comments.

### PR DESCRIPTION
**Pull request by julienfalque**
Related to #4913.

### CODE REVIEW

File: tests/DocBlock/AnnotationTest.php [hunk 1]
```diff
@@ -309,6 +309,30 @@ public function provideTypeParsingCases()
...
+            [
+                '/** @var Foo::A*',
+                ['Foo::A*'],
+            ],
+            [
+                '/** @var array<Foo::A*>|null',
+                ['array<Foo::A*>', 'null'],
+            ],
+            [
+                "/** @var null|true|false|1|1.5|'a'|'b'",
```
Comment (julienfalque) [inline tests/DocBlock/AnnotationTest.php:333]: @Nyholm I'm not familiar with Psalm: does it support all these types? Does it support double quoted strings?

File: tests/DocBlock/AnnotationTest.php [hunk 2]
```diff
@@ -309,6 +309,30 @@ public function provideTypeParsingCases()
                 '/** @var array<string|int, string>',
                 ['array<string|int, string>'],
             ],
+            [
+                '/** @var Foo::*',
```
Comment (Nyholm) [inline tests/DocBlock/AnnotationTest.php:313]: Also add `Foo::STATUS_*`

Comment (julienfalque) [reply tests/DocBlock/AnnotationTest.php:313]: Isn't `Foo::A*` below fine?

File: tests/DocBlock/AnnotationTest.php (hunk 1 above)
Comment (Nyholm) [reply tests/DocBlock/AnnotationTest.php:333]: Great question. I had to look it up. 

Yes, it does: https://psalm.dev/r/6564c220dc

File: tests/DocBlock/AnnotationTest.php (hunk 2 above)
Comment (Nyholm) [reply tests/DocBlock/AnnotationTest.php:313]: Oh, I missed that. Good!

File: tests/DocBlock/AnnotationTest.php (hunk 1 above)
Comment (Nyholm) [reply tests/DocBlock/AnnotationTest.php:333]: Yes, it all looks good. 

https://psalm.dev/r/66146836e6

Comment (julienfalque) [reply tests/DocBlock/AnnotationTest.php]: Cool, I updated the test case with a double quoted string, thanks for checking!

**Comment by julienfalque** [comment]
I've worked a bit more on this and made several improvements. Now this PR adds support for:
* spaces around `|`, e.g.: `A | B`
* class constants, e.g. `Foo::BAR_*`
* scalar constants, e.g. `'a' | 'b'`
* types containing a `-`, e.g. `class-string`
* intersection types, e.g. `A & B`

File: tests/DocBlock/AnnotationTest.php [hunk 3]
```diff
@@ -309,6 +309,62 @@ public function provideTypeParsingCases()
...
+            [
+                '/** @param int | "a" | A<B<C, D>, E<F::*|G[]>> $foo */',
+                ['int', '"a"', 'A<B<C, D>, E<F::*|G[]>>'],
+            ],
+            [
+                '/** @var class-string<Foo> */',
+                ['class-string<Foo>'],
+            ],
+            [
+                '/** @var A & B */',
```
Comment (Slamdunk) [inline tests/DocBlock/AnnotationTest.php:353]: Does this cover also spaceless case `A&B`?

Comment (julienfalque) [reply tests/DocBlock/AnnotationTest.php:357]: It does, I added a test case.

**Comment by Slamdunk** [comment]
> * spaces around `|`, e.g.: `A | B`
> * class constants, e.g. `Foo::BAR_*`
> * scalar constants, e.g. `'a' | 'b'`
> * types containing a `-`, e.g. `class-string`
> * intersection types, e.g. `A & B`
> * object-like arrays, e.g.: `array{a: bool, b?: string}`

I'm sorry @keradus to ping you, but [`v2.16.2 Yellow Bird`](https://github.com/FriendsOfPHP/PHP-CS-Fixer/releases/tag/v2.16.2) kinda broke almost every project I'm aware of, since these notations are now widespread due to static analysis tools.

Hope to see a release with this bugfix as soon as possible :cry:

**Comment by julienfalque** [comment]
@Slamdunk Note that the actual fix for the unexpected removal of annotations is #4915. This PR will rather unlock enhancements in e.g. `phpdoc_to_return_type`.

**Comment by Slamdunk** [comment]
Yup, sorry

**Comment by kubawerlos** [comment]
@julienfalque can we add tests from https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/5049 here?

**Comment by julienfalque** [comment]
> after gitter talk

Didn't get a notification for this review being dismissed. What was the talk about?

> @julienfalque can we add tests from #5049 here?

Sure thing. Are you allowed to push on my branch? Otherwise I'll cherry-pick your commit.

**Comment by kubawerlos** [comment]
@julienfalque No, I cannot, please cherry-pick.

**Comment by julienfalque** [comment]
@kubawerlos I cherry-picked your work and changed it a bit to solve conflicts and add support for [`Closure` syntax](https://psalm.dev/docs/annotating_code/type_syntax/callable_types/).

**Comment by GrahamCampbell** [comment]
Is there support for brackets in types. Like `array{0: (string|null|(callable(): void)), 1: string`. Or `(callable(Foo): string)|null`.

**Comment by julienfalque** [comment]
I don't think this is supported but since it might be more complicated than adding support for a specific type syntax, I would suggest merging this PR first.

**Comment by kubawerlos** [comment]
@Slamdunk @Nyholm @SpacePossum @keradus is anything stops us from having this one merged?

**Review by Slamdunk — APPROVED** [review]

**Comment by SpacePossum** [comment]
shouldn't this pass as well as types never have spacing attached?

```php
            [
                '/** @param array<  int   , callable  (  string  )  :   bool  > $function',
                ['array<int, callable(string): bool>'],
            ],
```

**Comment by julienfalque** [comment]
@SpacePossum You mean to add support for whitespace anywhere?

**Comment by SpacePossum** [comment]
indeed, I don't see why the spacing should when extracting the type info

**Comment by julienfalque** [comment]
Using whitespaces is not a widespread practice but it should not be complicated to add support for it. I'll try to do it shortly.

**Comment by SpacePossum** [comment]
:+1:   it will saves the trouble of figuring out all the possible priority issues with fixers that clean up the spacing

**Comment by julienfalque** [comment]
Test case added. Though whitespaces are not stripped from the result because the type is parsed as a whole.

**Comment by SpacePossum** [comment]
Is the "type" `"   int  "` not exactly the same as "type" `"int"` ? Especially as the type `"   int  "` can never exist within PHP, same with class names that cannot contain pending or trialing spacing.  I'm not sure why we would extract these differently but maybe I'm missing the use case here.

**Comment by julienfalque** [comment]
Not sure what you mean. The regex just extracts the type definition as-is, without changing anything to it. Those whitespaces have no special meaning, they are not part of a type, e.g. there is no `" int "` but `"int"` + whitespaces.

Removing the whitespaces would prevent, e.g. a fixer to know the original syntax in order to change it.

**Comment by SpacePossum** [comment]
@kubawerlos @julienfalque RTM than? can someone run this code on the some other code bases like SF/Twig/Doctrine for example to see if all goes well?

**Comment by SpacePossum** [comment]
and one more request, can some look into https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/5095/files to see if the issue there is no longer present in this PR?

**Comment by kubawerlos** [comment]
> and one more request, can some look into https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/5095/files to see if the issue there is no longer present in this PR?

Can't we merge this and update the other one and we will all see?

**Comment by SpacePossum** [comment]
I prefer to resolve both PR's here as 5095 hasn't got any comments what so ever and I don't want it to sit there forever

**Comment by julienfalque** [comment]
@SpacePossum I tried applying changes from #5095 and all tests were green. I approved #5095, we can merge them separately IMO.

**Comment by keradus** [comment]
can you rebase, @julienfalque ?

**Comment by julienfalque** [comment]
Done.

File: src/DocBlock/TypeExpression.php [hunk 4]
```diff
@@ -25,31 +25,74 @@ final class TypeExpression
...
-            (?<simple>
-                [@$?]?[\\\\\w]+
-            )
-            |
-            (?<generic>
-                (?&simple)
-                <
-                    (?:(?&types),\s*)?(?:(?&types)|(?&generic))
-                >
+
```
Comment (keradus) [inline src/DocBlock/TypeExpression.php:28]:
```suggestion
```

**Comment by keradus** [comment]
Thank you @julienfalque.

**Comment by Slamdunk** [comment]
@julienfalque do you think this could be behind https://github.com/FriendsOfPHP/PHP-CS-Fixer/issues/5836?

**Pull request merged**